fix!: drop support for Python 2.7 / 3.5

tseaver opened the pull request:
- Drop `six`.
- Drop other Python 2.7-specific workarounds (`collections_abc`, etc.)
- Drop `u"` prefixes for text strings.
- Drop `pytz` (per googleapis/python-api-core#73).

Release-As: 2.0.0b1

Closes googleapis/python-api-core#210.

tseaver committed:
fix!: drop support for Python 2.7 and 3.5

Release-As: 2.0.0b1

tseaver committed:
chore: drop 'six' module

tseaver committed:
chore: drop 'u"' prefixes for text

tseaver committed:
chore: remove other Python 2.7 workarounds

tseaver committed:
chore: drop use of 'pytz'

Closes #74.

plamut commented on README.rst:
```diff
@@ -16,8 +16,4 @@ common helpers used by all Google API clients. For more information, see the
 -------------------------
-Python >= 3.5
-
-Deprecated Python Versions
---------------------------
-Python == 2.7. Python 2.7 support will be removed on January 1, 2020.
```
IIRC this section should be retained, if only to explain what the last Python 2.7 compatible version is.

BigQuery example: https://googleapis.dev/python/bigquery/latest/index.html#unsupported-python-versions
Pub/Sub docs example: https://googleapis.dev/python/pubsub/latest/index.html#deprecated-python-versions

tseaver replied:
fe7d454

plamut commented on google/api_core/version.py:
```diff
@@ -12,4 +12,4 @@
 # limitations under the License.
 
-__version__ = "1.30.0"
+__version__ = "2.0.0-b1"
```
Is this a new process, i.e. having a preliminary release instead of immediately switching to 2.0.0? Also, should this be beta or an RC?

FWIW, the approach itself does seem sensible to me.

tseaver replied:
As a semver-major / breaking change, I thought it was reasonable to do a short beta.  I could be persuaded that it should be an RC, but I want folks to consider carefully whether they install it in production until we've gotten feedback that all is OK.

plamut commented on google/api_core/future/base.py:
```diff
@@ -16,11 +16,8 @@
 
 import abc
 
-import six
```
I noticed that we still have a `six` pin in `testing/constraints-3.6.txt`, let's remove that, too.

tseaver replied:
6e21b01

plamut commented on tests/unit/test_exceptions.py:
```diff
@@ -138,15 +138,15 @@ def test_from_http_response_bad_json_content():
 
     assert isinstance(exception, exceptions.NotFound)
     assert exception.code == http.client.NOT_FOUND
-    assert exception.message == u"POST https://example.com/: \u2019 message"
```
I found a few more of these in `docs/conf.py`, let's remove them there, too.

tseaver replied:
Templated file -- I'd rather fix that upstream in `synthtool`.

tseaver replied:
https://github.com/googleapis/synthtool/pull/1119

plamut reviewed (commented):
Generally looks good, just spotted a few legacy bits that were not cleaned yet.

There are also a few remaining references to _"2.7"_ and _"Python 2"_ (comments, workarounds, etc.) that can be removed. And maybe the [workaround](https://github.com/googleapis/python-api-core/blob/11032cf08ecc16dd252a6cda8b33b0b28ec4f4ba/google/api_core/bidi.py#L640-L652) for the lack of `Condition.wait_for()` method in Python 2.7.

---

**Edit:** Oh, and we also use the `mock` shim in unit tests (installed in noxfile's default session), we can now switch to importing the standard library's `unittest.mock`.

**Edit 2:** We can also get rid of several `def next(self)` methods and replace them with `__next__(self)`.

tseaver committed:
Merge branch 'master' into 210-drop-python-2.7

tseaver committed:
tests: scrub more 2.7 / six constraints

tseaver committed:
docs: add section on unspported Python versions to README

tseaver committed:
chore: drop workaround for Python 2.7

tseaver committed:
chore: remove 'general_helpers.wraps'

Except for a backward-compatibility import, 'functools.wraps' does
everything wee need on Python >= 3.6.

tseaver committed:
chore: note URL for follow-up issue

tseaver committed:
chore: remove Python 2-specific kwargs hack

tseaver committed:
chore: drop Python2-specific alias

tseaver committed:
chore: drop Python2-specific aliases

tseaver committed:
chore: drop Python2-specific aliases

tseaver commented:
@plamut Thanks for the review

> There are also a few remaining references to "2.7" and "Python 2" (comments, workarounds, etc.) that can be removed.

I'll leave the bit in `docs/auth.rst` for now.  I'm working through a couple of others.

> And maybe the workaround for the lack of Condition.wait_for() method in Python 2.7.

I added #211 to track that change:  I didn't want to think that hard in the middle of a big janitorial cleanup.

> Edit: Oh, and we also use the `mock` shim in unit tests (installed in noxfile's default session), we can now switch
> to importing the standard library's unittest.mock.

We can't replace `import mock` wholesale with `from unittest import mock`, because the stdlib version doesn't have 'mock.AsyncMock` before Python 3.8.

> Edit 2: We can also get rid of several def next(self) methods and replace them with __next__(self).

4165633, ad7b40b, dbc234f.

plamut reviewed (approved):
LGTM, thanks for the additional cleanup.

It's totally fine to defer some of the less trivial (and thus risky) changes to a separate PR, esp. with issues to track them.

parthea commented on setup.py:
```diff
@@ -35,8 +35,6 @@
     "requests >= 2.18.0, < 3.0.0dev",
     "setuptools >= 40.3.0",
     "packaging >= 14.3",
```
Can we remove packaging also? See #215

tseaver replied:
72d175b

tseaver commented:
Marked as `do not merge` pending discussion over the versioning policy next week with @busunkim96, @crwilcox, et al.

parthea commented on owlbot.py:
```diff
@@ -22,8 +22,15 @@
+    "setup.cfg",  # pytype
+    ".flake8",  # flake8-import-order, layout
+    ".coveragerc",  # layout
+    "CONTRIBUTING.rst",  # no systests
```
Can we keep `CONTRIBUTING.rst` and just update the SHA in `.OwlBot.lock.yaml` to `b6169fc6a5207b11800a7c002d0c5c2bc6d82697185ca12e666f44031468cfcd`?

tseaver replied:
There should be no "System Tests" section at all in `CONTRIBUTING.rst` for this repository -- it has none.  I don't know that the generator is configurable enough to allow that change, and I'd really like not to have to add the monster `s.replace` required to remove it.

parthea replied:
Makes sense.

tseaver committed:
Merge branch 'master' into 210-drop-python-2.7

tseaver committed:
Merge branch 'master' into 210-drop-python-2.7

tseaver committed:
chore: remove 'packaging' dependency

Closes #215.

tseaver committed:
Merge branch 'master' into 210-drop-python-2.7

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/master/packages/owl-bot/README.md

tseaver committed:
chore: switch to microgenerator to avoid 2.7 reappearing

tseaver committed:
chore: expand range to allow 'google-auth' 2.x versions

tseaver committed:
Merge branch 'master' into 210-drop-python-2.7

tseaver committed:
Merge branch 'master' into 210-drop-python-2.7

tseaver committed:
docs: sync newest version w/ Python 2.7 support

tseaver committed:
chore: drop comment referring to Python 2.7 / pytypes

tseaver committed:
chore: remove mentions of Python 2.7 from docs

Except for the README section documentin the last version supporting it.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/master/packages/owl-bot/README.md

tseaver commented:
Ugh, OwlBot is restoring the `2.7` reference in the templated `scripts/readme-gen/templates/install_deps.tmpl.rst'.  See https://github.com/googleapis/synthtool/pull/1173.
